R/extract_data.R

```diff
@@ -0,0 +1,245 @@
+  if (table == "contexts") {
+
+    table_tmp <- database$contexts %>% 
+      dplyr::filter(stringr::str_detect(database$contexts[[col]], col_value)) %>%
+      dplyr::distinct(link_id)
+    
+    table <- as.vector(table_tmp[[1]])
+
+    if (length(table) == 0) {
+      table <-  as.vector("treatment_context_id")
+    }
+  
+  }
+  
+  for (i in seq_along(1:length(table))) {
+    
+      tables_tmp <- tables
+      
+      # chose columns to select, ensuring "value" isn't among the columns, since it has a different meaning for each table
+      columns_to_select <- intersect(setdiff(names(database$traits), "value"), names(database[[table[[i]]]]))
+      
+      # Trim reference table i.e. locations with pattern matching
+      #found_indicies <- database[[table[[i]]]][[col]] |> stringr::str_which(pattern = stringr::regex(col_value, ignore_case = TRUE))
+      
+      indicies_tmp <- purrr::map(col_value, ~{
+        stringr::str_which(database[[table[[i]]]][[col]], 
+                           pattern = stringr::regex(.x, ignore_case = TRUE))
+      })
+      
+      found_indicies <- purrr::reduce(indicies_tmp, union)
+  
+      # Trim traits, based on the columns identified as being common between the traits table and target table
+      cc_traits <- database[[table[[i]]]] |>
+        dplyr::slice(found_indicies) |>
+        dplyr::select(tidyselect::all_of(columns_to_select)) |> 
+        dplyr::distinct()
+      
+      # Filtering join
+      ## It will quite literally cookie cutting the traits table if the columns match what is in cc_traits
+      ret_tmp[["traits"]] <- database[["traits"]]|> 
+        dplyr::semi_join(cc_traits, by = columns_to_select)
+      
+      # Use same filtering join to trim excluded data
+      ret_tmp[["excluded_data"]] <- database[["excluded_data"]]  |> 
+        dplyr::semi_join(cc_traits, by = columns_to_select)
+      
+      
+      for (j in seq_along(tables_tmp$tables_to_cut)) {
+      
+        cut_traits <- ret_tmp[["traits"]] |> 
+          dplyr::select(get(tables_tmp$cookie_cutters[[j]])) |> 
+          dplyr::distinct() 
+        
+        cut_traits <- cut_traits |>
+          dplyr::filter(dplyr::if_all(tidyselect::everything(), ~ !is.na(.)))
+        
+        cut_table <- eval(parse(text = tables_tmp$tables_complete_path[[j]])) |>
+          dplyr::semi_join(cut_traits, by = get(tables_tmp$cookie_cutters[[j]])) %>%
+          dplyr::rename(link_vals = tidyselect::contains("context_id"))
+        
+        assign(paste0("ret_tmp[[\"", tables_tmp$tables_to_cut[[j]], "\"]]"), cut_table )
+      
+      }
+      
+      # Bind together rows extracted from each table
+      ## This step is required when the `context` table is queried as it is split into 5 tables and each needs to be checked.
+      ret[["traits"]] <- ret[["traits"]] %>% dplyr::bind_rows(ret_tmp[["traits"]]) %>% dplyr::distinct()
```

Comment:
The backticking is still very much a mystery to me and why it applies to only some tables and not others. Do u have a good explainer @dfalster ?

Reply:
I had it written out explicitly because I kept getting errors ... and I continue to, for instance:
```
>         ret[[v]] <- ret_tmp[[v]] %>%
+           dplyr::bind_rows(`ret_tmp[[v]]`) %>%
+           dplyr::distinct()
Error: object 'ret_tmp[[v]]' not found
```
But:
```

> ret_tmp[[v]]
# A tibble: 1,540 × 26
   dataset_id taxon_name observation_id trait_name value unit  entity_type value_type basis_of_value replicates basis_of_record life_stage population_id individual_id repeat_measurements_id
   <chr>      <chr>      <chr>          <chr>      <chr> <chr> <chr>       <chr>      <chr>          <chr>      <chr>           <chr>      <chr>         <chr>         <chr>                 
 1 Campbell_… Acacia fa… 001            bud_bank_… basa… NA    population  mode       expert_score   NA         field           adult      01            NA            NA                    
 2 Campbell_… Acacia fa… 001            resprouti… resp… NA    population  mode       expert_score   NA         field           adult      01            NA            NA                    
 3 Campbell_… Acacia fa… 001            seedbank_… soil… NA    population  mode       expert_score   NA         field           adult      01            NA            NA

Reply:
Ah yes, the back tick is a mistake. 

```
for(v %in% c("xxx", ".yyy", ...)) {
  ret[[v]] <- ret[[v]] %>% dplyr::bind_rows(ret_tmp[[v]]) %>% dplyr::distinct()
}
```

Reply:
I spent a long time working out how to recode everything in this function without the backticks. Which made a lot of pieces much easier.